Multiplication and subtraction methods

julianmateu opened the pull request:
On this PR I added methods and their respective tests for:
1. Subtract one quaternion from another.
2. Multiply one quaternion by a scalar.
3. Multiply two quaternions and find their Hamilton product.

@basicNew

julianmateu committed:
Add methods to subtract and multiply

-Add `subtract()` and `multiply()` methods in Quaternion.java, but no tests.

julianmateu committed:
Add method to multiply by scalar and tests

-Added a method to multiply by scalar.
-Added tests for subtract, and multiply methods.

The multiply by quaternion method has problems with tolerance.

julianmateu committed:
Corrected mistake, still have error with tolerance

-Corrected and extra ")" int TestQuaternion.java, but still have a tolerance problem in the test with inverse.

julianmateu committed:
Corrected the tolerance error.

Add a new function `assertEqualsQuaternion(Quaternion,Quaternion,delta)`.

julianmateu committed:
Corrected format

basicNew commented on src/Quaternion.java:
```diff
@@ -20,75 +24,151 @@ public Quaternion(double r, double i, double j, double k) {
-    public double getR() {return this.r;}
+    public Quaternion add(Quaternion addend) {
+        return new Quaternion(
+                this.r + addend.getR(),
```
use getters to encapsulate yourself form the implementation

basicNew commented on src/Quaternion.java:
```diff
@@ -20,75 +24,151 @@ public Quaternion(double r, double i, double j, double k) {
-    public double getI() {return this.i;}
+    public Quaternion subtract(Quaternion subtrahend) {
+        return new Quaternion(
+                this.r - subtrahend.getR(),
```
same as above

basicNew commented on src/Quaternion.java:
```diff
@@ -20,75 +24,151 @@ public Quaternion(double r, double i, double j, double k) {
-    public double getK() {return this.k;}
+    public boolean equals(Quaternion compared) {
+
+        return this.r == compared.getR() && this.i == compared.getI() && this.j ==
```
same as above

basicNew commented on src/Quaternion.java:
```diff
@@ -20,75 +24,151 @@ public Quaternion(double r, double i, double j, double k) {
-    public void setR(double r) {this.r = r;}
+    @Override
+    public String toString() {
+        return "( " + this.r + " ; " + this.i + " ; " + this.j + " ; " + this.k + " )";
```
same as above

basicNew commented on src/Quaternion.java:
```diff
@@ -20,75 +24,151 @@ public Quaternion(double r, double i, double j, double k) {
-    @Override
-    public String toString(){
-        return "( " + this.r + " ; " + this.i + " ; " + this.j + " ; " + this.k + " )";
+    public void setK(double k) {
```
Do we need public setters?

basicNew commented on src/TestQuaternion.java:
```diff
@@ -1,49 +1,183 @@
+                new Quaternion(0, 0, 0, 0)); // Multiplying by right zero element.
+        assertEquals(new Quaternion(0, 0, 0, 0).multiply(new Quaternion(1, 1, 1, 1)),
+                new Quaternion(0, 0, 0, 0)); // Multiplying by left zero element.
+        assertEqualsQuaternion(new Quaternion(1, 2, 3, 4).multiply(new Quaternion(
```
you may want to use tempoaral variables in these cases, as the test is a bit hard to read

basicNew commented:
Just a few comments, otherwise looks good @julianmateu ! Let me know when those are addressed and I'll merge

julianmateu committed:
Correct mistakes in PR

-Use getters in all Quaternion methods.
-Make the setters private.
-Add auxiliary variables in the multiplication test cases to make them easier to read.

julianmateu commented:
@basicNew I think I've corrected the mistakes now.
